### PR TITLE
Clone profiles

### PR DESCRIPTION
**Pull request by cconard96**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -

Enable cloning for profiles. This will be useful for admins by letting them start from an existing profile rather than going through every tab and manually adding all the different rights they want.

### CODE REVIEW

**Comment by trasher** [comment]
This need unit tests

File: tests/functionnal/Profile.php [hunk 1]
```diff
@@ -194,4 +194,53 @@ public function testSwitchingInterface()
...
+        $profile = new \Profile();
+        $this->boolean($profile->getFromDB(3))->isTrue();
+
+        // Clone it
+        $cloned_profile = new \Profile();
+        $clone_profiles_id = $profile->clone([
+            'name' => __FUNCTION__,
+        ]);
+        $this->integer($clone_profiles_id)->isGreaterThan(0);
+        $cloned_profile->getFromDB($clone_profiles_id);
```
Comment (trasher) [inline tests/functionnal/Profile.php:212]:
```suggestion
        $this->boolean($cloned_profile->getFromDB($clone_profiles_id))->isTrue();
```

File: tests/functionnal/Profile.php [hunk 2]
```diff
@@ -194,4 +194,53 @@ public function testSwitchingInterface()
...
+        ]);
+        // Check that all rights with profiles_id 3 exist with the clone ID as well
+        $rights = [
+            3 => [],
+            $clone_profiles_id => [],
+        ];
+        foreach ($rights_iterator as $right) {
+            $rights[$right['profiles_id']][$right['name']] = $right['rights'];
+        }
+        $this->integer(count($rights[$clone_profiles_id]))->isGreaterThanOrEqualTo(count($rights[3]));
```
Comment (trasher) [inline tests/functionnal/Profile.php:240]: I guess the rights count should be the same, no?
```suggestion
        $this->integer(count($rights[$clone_profiles_id]))->isEqualTo(count($rights[3]));
```

File: src/ProfileRight.php [hunk 3]
```diff
@@ -45,6 +45,34 @@ class ProfileRight extends CommonDBChild
     public static $items_id = 'profiles_id'; // Field name
     public $dohistory       = true;
 
+    public function clone(array $override_input = [], bool $history = true)
```
Comment (cedric-anne) [inline src/ProfileRight.php:48]: Could you explain why you overrided the `clone` mehod ?

File: src/ProfileRight.php [hunk 4]
```diff
@@ -45,6 +45,34 @@ class ProfileRight extends CommonDBChild
...
+    {
+        global $DB;
+
+        if ($DB->isSlave()) {
+            return false;
+        }
+        $new_item = new static();
+        $input = Toolbox::addslashes_deep($this->fields);
```
Comment (cedric-anne) [inline src/ProfileRight.php:56]:
```suggestion
        $input = Sanitizer::dbEscapeRecursive($this->fields);
```

File: src/ProfileRight.php (hunk 3 above)
Comment (cconard96) [reply src/ProfileRight.php:48]: The default implementation uses an `add` call to create the cloned relations. In the case of profiles, the the rights are added when the Profile is created with values of 0 so we have to do an update or insert to properly "clone" the rights from the original Profile.

Comment (cedric-anne) [reply src/ProfileRight.php:50]: Could you add a comment in code ?

**Review by cedric-anne — APPROVED** [review]

**Review by trasher — APPROVED** [review]

**Review by orthagh — APPROVED** [review]

**Comment by anibaldiniz** [comment]
Is it possible in the current version to clone profiles?

**Pull request merged**